text/0477-policy-validation.md

```diff
@@ -0,0 +1,553 @@
+some template is non-compliant, you can get this feedback immediately, every time you synthesize the cloud assembly. If
+your application is not compliant, the CLI will report error messages, how to fix the issues (depending on the
+integrated tool) and the offending construct.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+One of the recurring complaints about the CDK from large enterprise customers is that it generates many resources
+automatically. While this is the original intent of the CDK, it also creates a challenge from a security and compliance
+standpoint; the generated resource definitions might not satisfy the rules defined by the central operations teams,
+which, at best, creates delays — as a result of development rework if the issues are caught before deployment — and, at
+worst, may expose the customer to attacks or lawsuits.
+
+By integrating policy-as-code tools in the synthesis flow, we give the confidence that these customers need to adopt the
+CDK at a larger scale.
+
+### What is the technical solution (design) of this feature?
+
+See [Appendix A - High level design](https://quip-amazon.com/lDPqAisXjVP6#temp:C:CaFafef0ee17e804c3bb00434ff5).
+
+### Is this a breaking change?
+
+No.
+
+### What alternative solutions did you consider?
+
+We have considered two alternative solutions:
+
+#### Validation in the CLI
+
+With this solution, the validation would be done in the CLI, rather than in the framework. With this approach, users
+would configure the validation plugins using a config file (cdk.json would be a natural candidate):
+
+```json
+"validations": {
+    "package": "cdk-cfn-guard",
+    "class": "CfnFuardValidation",
+    "version": "1.2.3"
+}
+```
+
+The CLI would read this file, load the appropriate plugins either from the filesystem or directly from NPM, and call
+them to get a validation report.
+
+We eventually discarded this option for the following reasons:
+
+* All plugins would have to be written in TypeScript/JavaScript for them to be consumable by the CLI.
+* The CLI lacks contextual information that would be useful for tracing the problem back to where it was defined.
+* If an application is synthesized without using the CDK CLI, it’s not subject to policy validation.
+
+#### CloudFormation hooks integration
+
+Instead of integrating with each policy tool individually, we have considered the option of developing a single
+mechanism, that would integrate with CloudFormation hooks. The idea was to invoke the Lambda function underlying a
+CloudFormation hook and interpret its results in a similar way CloudFormation, that is, by failing synthesis if the
+function reported a validation failure.
+
+However, the Lambda functions invoked by CloudFormation hooks are not deployed to the customer’s account, but to an
+account owned by AWS. This makes this solution not feasible.
+
+### What are the drawbacks of this solution?
```

Comment:
Another drawback I'd like to point out:

- If an organization wants validations to apply to all languages in the company, they have to setup a jsii publishing pipeline. But if the validation is invoked from the CLI, than all that is needed is one node package.